### PR TITLE
Image upload backend implementation

### PR DESCRIPTION
**Pull request by Tirokk**
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-20T18:38:11Z" title="Monday, May 20th 2019, 8:38:11 pm +02:00">May 20, 2019</time>_
_Merged <time datetime="2019-05-27T12:28:00Z" title="Monday, May 27th 2019, 2:28:00 pm +02:00">May 27, 2019</time>_
---

This is a curl request from a folder where I have a file `avatar.jpg`:

```
curl localhost:4000/ \
  -F operations='{ "query": "mutation($id: ID!, $file: Upload) { UpdateUser(id: $id, avatarUpload: $file) { id name avatar } }", "variables": { "id": "u3", "file": null } }' \
  -F map='{ "0": ["variables.file"] }' \
  -F 0=@avatar.jpg
```

It uploads the avatar to the public folder. 

Same with the posts:
```
curl localhost:4000/ \
                         -H "Authorization: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidXNlciIsIm5hbWUiOiJKZW5ueSBSb3N0b2NrIiwiZGlzYWJsZWQiOmZhbHNlLCJhdmF0YXIiOiIvdXBsb2Fkcy8xNTU4MzkxOTM0NzcwLWF2YXRhciIsImlkIjoidTMiLCJlbWFpbCI6InVzZXJAZXhhbXBsZS5vcmciLCJzbHVnIjoiamVubnktcm9zdG9jayIsImlhdCI6MTU1ODM5NTc1OSwiZXhwIjoxNjQ0Nzk1NzU5LCJhdWQiOiJodHRwOi8vbG9jYWxob3N0OjMwMDAiLCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjQwMDAiLCJzdWIiOiJ1MyJ9.IAjxXWs7bRMZyCV8c56oDoPRU3vfUED1vwQc3lwPCRA" \
                         -F operations='{ "query": "mutation($id: ID!, $file: Upload) { UpdatePost(id: $id, imageUpload: $file) { id title image } }", "variables": { "id": "9d5b2cab-c447-4e7e-8a57-e58daaba3306", "file": null } }' \
                         -F map='{ "0": ["variables.file"] }' \
                         -F 0=@avatar.jpg
```
Note that you must be authenticated and it must be your own post.

### Issues

- [X] fixes: #638

### Checklist

- [X] None

### How2Test

- [X] None

### Todo
* [x] Write tests
* [ ] Change the `user.avatar` url and send it to `/api` in the frontend

### CODE REVIEW

File: backend/src/resolvers/users.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+export default {
+  Mutation: {
+    UpdateUser: async (object, params, context, resolveInfo) => {
+      params = await fileUpload(params, { file: 'avatarUpload', url: 'avatar' })
+      return neo4jgraphql(object, params, context, resolveInfo, false)
+    },
+    CreateUser: async (object, params, context, resolveInfo) => {
+      params = await fileUpload(params, { file: 'avatarUpload', url: 'avatar' })
+      return neo4jgraphql(object, params, context, resolveInfo, false)
+    },
```
Comment (Tirokk) [inline backend/src/resolvers/users.js:13]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-05-21T23:00:36Z" title="Wednesday, May 22nd 2019, 1:00:36 am +02:00">May 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,15 @@
+import { neo4jgraphql } from 'neo4j-graphql-js'
+import fileUpload from './fileUpload'
+
+export default {
+  Mutation: {
+    UpdateUser: async (object, params, context, resolveInfo) => {
+      params = await fileUpload(params, { file: 'avatarUpload', url: 'avatar' })
+      return neo4jgraphql(object, params, context, resolveInfo, false)
+    },
+    CreateUser: async (object, params, context, resolveInfo) => {
+      params = await fileUpload(params, { file: 'avatarUpload', url: 'avatar' })
+      return neo4jgraphql(object, params, context, resolveInfo, false)
+    }
```


Love :heart_eyes_cat: how clean this code is!!

**Comment by Tirokk** [comment]
> [codecov[bot]](https://github.com/apps/codecov) **Authored by [codecov[bot]](https://github.com/apps/codecov)**
_<time datetime="2019-05-23T19:24:35Z" title="Thursday, May 23rd 2019, 9:24:35 pm +02:00">May 23, 2019</time>_

---


# [Codecov](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636?src=pr&el=h1) Report
> Merging [#636](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636?src=pr&el=desc) into [master](https://codecov.io/gh/Human-Connection/Human-Connection/commit/728715818112d9cceacffb8a8aea1b2207e49870?src=pr&el=desc) will **increase** coverage by `0.31%`.
> The diff coverage is `52.38%`.

[Impacted file tree graph](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #636      +/-   ##
==========================================
+ Coverage    20.2%   20.51%   +0.31%     
==========================================
  Files         157      159       +2     
  Lines        2208     2228      +20     
  Branches      105      105              
==========================================
+ Hits          446      457      +11     
- Misses       1738     1747       +9     
  Partials       24       24
```


| [Impacted Files](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [backend/src/graphql-schema.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvZ3JhcGhxbC1zY2hlbWEuanM=) | `0% <ø> (ø)` | :arrow_up: |
| [backend/src/middleware/notifications/index.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvbWlkZGxld2FyZS9ub3RpZmljYXRpb25zL2luZGV4Lmpz) | `0% <ø> (ø)` | :arrow_up: |
| [backend/src/middleware/validation/index.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvbWlkZGxld2FyZS92YWxpZGF0aW9uL2luZGV4Lmpz) | `0% <0%> (ø)` | :arrow_up: |
| [backend/src/resolvers/users.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvcmVzb2x2ZXJzL3VzZXJzLmpz) | `0% <0%> (ø)` | |
| [backend/src/resolvers/posts.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvcmVzb2x2ZXJzL3Bvc3RzLmpz) | `0% <0%> (ø)` | :arrow_up: |
| [...d/src/middleware/notifications/extractIds/index.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvbWlkZGxld2FyZS9ub3RpZmljYXRpb25zL2V4dHJhY3RJZHMvaW5kZXguanM=) | `100% <100%> (ø)` | |
| [backend/src/resolvers/fileUpload/index.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/636/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvcmVzb2x2ZXJzL2ZpbGVVcGxvYWQvaW5kZXguanM=) | `83.33% <83.33%> (ø)` | |

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-05-27T12:27:53Z" title="Monday, May 27th 2019, 2:27:53 pm +02:00">May 27, 2019</time>_

---


Merging as we discussed

**Review by Mogge — APPROVED** [review]
this is really great @roschaefer, I have tested this from the frontend as well, and it's working!!
Great job adding so many tests and writing such clean code!!

**Pull request closed without merging**